sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala

```diff
@@ -412,6 +412,11 @@ case class CatalogRelation(
   assert(tableMeta.partitionSchema.sameType(partitionCols.toStructType))
   assert(tableMeta.dataSchema.sameType(dataCols.toStructType))
 
+  // If table schema is empty, Spark will infer the schema at runtime, so we should mark this
+  // relation as unresolved and wait it to be replaced by relation with actual schema, before
+  // resolving parent plans.
+  override lazy val resolved: Boolean = tableMeta.schema.nonEmpty
```

Comment:
https://github.com/apache/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala#L1894-L1923

We have a test case. I also saw multiple related JIRAs. The users complain the table with zero column does not have a correct number of rows after `EXCEPT` and `INTERSECT`...